rfcs/text/0014_notification_service.md

```diff
@@ -0,0 +1,373 @@
+
+*Repository* interface:
+- `create(NotificationEvent)`- creates a notification in the Notification system. 
+- `get({ search_after?: number; size?: number; type_id?: string[] })` - retrieves last notifications for the current user.
+Might be paginated using `search_after` filed.
+- `markAsRead({ notification_id: string[] })` - mark notifications with given ids as read.
+- `markAsUnread({ notification_id: string[] })` - mark notifications with given ids as unread.
+- `markAllAsRead()` - mark all the notifications as read.
+- `markAsPin({ notification_id: string[] })` - mark notifications with given ids as pinned.
+- `markAsUnpin({ notification_id: string[] })` - mark notifications with given ids as unpinned.
+- `configure()` - updates user-specific notification settings.
+
+#### LocalRepository
+![image](../images/0014/local_repository.png)
+
+##### Notification model
+Notification model is responsible for:
+- Applying policies to an incoming notification. NS must check whether a recipient unsubscribed from a source before processing to delivery.
+- Enhancing an incoming notification with necessary data: 
+    - `expired_at` if not specified. Kibana allows administrators to set how long notifications will be stored. Falls back to 30 - 90 days be default. 
+- Writing new incoming messages to the Notification storage.
+
+Processing a large number of incoming notifications might slow down Kibana and create a disproportionately high load on some Kibana instances.
+The use of a queue allows us to make the handling of incoming notifications asynchronously.
+Async mechanism lets Kibana to cope with spikes in the number of messages created. The queue implementation is based on the Elasticsearch server. 
+
+The logic is run as a background task. The usage of the single background task:
+- reduces the possibility of read- and write-conflicts
+- allows re-using task_manager built-in features: scheduling, retry, monitoring
+
+If Kibana processes notifications in asynchronous batches, we will need a prioritization mechanism to be sure events
+with a higher severity are delivered first without a significant delay. The prioritization mechanism is based on
+priority property value: the notification with the higher value is processed earlier to provide an experience as close to real-time as possible.
+
+##### Notification storage
+
+NS storage keeps different notification statuses for every user. Most Kibana deployments have [from 20 to 90 users](https://elastic.slack.com/archives/C9097ABGC/p1597162974399900),
+although there are some odd instances with >100k users. So we consider two options at the moment (see the [workflow](https://docs.google.com/document/d/18s-BTHog_oPaQKf871gzuhxkah4pK1GYJM9DCLceaLo/edit#)):
+- Store notification state separately from a notification. State for new notifications created for every user when they change the state.
+```typescript
+interface  NotificationState {
```

Comment:
Will we want audit logging for this state? Would administrators find it useful to know that a specific user read a specific notification?

Reply:
I'm not aware of such a requirement at the moment. @alexfrancoeur

Reply:
Let's say the answer is "no" for the MVP and we can think about it more when read-state moves server-side. I don't know enough about our audit logging myself, so feel free to speak up if anyone disagrees.

If read-state ends up being saved-object backed, will we kindof get this for free with audit logging or are system-managed saved objects exempt?

Reply:
> If read-state ends up being saved-object backed, will we kindof get this for free with audit logging or are system-managed saved objects exempt?

That's correct, we will have a record that someone updates SO (`saved_object_update` event, the [full list of events](https://github.com/elastic/kibana/blob/099e9bc50fafa89301d438ae7112330da5ab421c/docs/user/security/audit-logging.asciidoc)). But we will need to add integration if we want to provide a more meaningful event on notification state change: `notification_read`, `notification_unread`, etc.